dataflux_pytorch/benchmark/checkpointing/simulated/README.md

```diff
@@ -0,0 +1,115 @@
+# Benchmarking PyTorch Lightning Checkpoints with Google Cloud Storage
+
+*   This benchmarking script allows you to run and benchmark the performance of the [Fully Sharded Data Parallel (FSDP)](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.strategies.FSDPStrategy.html) save/load function in order to save and restore checkpoints from GCS. 
+
+*   This is a simulated version of the code that runs on CPU to test the performance of GCS as a backing store for AI/ML workloads without GPU.
+
+*   The script generates [state_dict](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html#what-is-a-state-dict-in-pytorch) that represents the state of the model. Each node then saves a part of the state_dict to GCS (the state_dict is divided among the nodes, such that the size of each shard is the same). During restore, each node reads its shard from GCS in parallel. The `.metadata` file, which is written and read by the Rank_0 (coordinator) node, contains information about which shard each node should read.
+
+*   The official implementation of FSDP in PyTorch Lightning for save_checkpoint can be found [here]
+(https://github.com/Lightning-AI/pytorch-lightning/blob/3627c5bfac704d44c0d055a2cdf6f3f9e3f9e8c1/src/lightning/fabric/strategies/fsdp.py#L419)
+*   The official implementation of FSDP in PyTorch Lightning for load_checkpoint can be found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/3627c5bfac704d44c0d055a2cdf6f3f9e3f9e8c1/src/lightning/fabric/strategies/fsdp.py#L519)
+
+*   While the official implementation of both `save_checkpoint` and `load_checkpoint` does many things beyond just saving/loading the state_dict, these additional steps have been omitted for the following reasons: 
+*   *   We wanted to demonstrate how GCS will scale with increasing model/checkpoint sizes without needing an actual GPU.
+*   *   These additional steps do not interact with GCS.
+
+*   If a user wishes to run an end to end FSDP implementation on an actual GPU, then please look at the example under the [multinode](https://github.com/GoogleCloudPlatform/dataflux-pytorch/tree/main/demo/lightning/checkpoint/multinode) directory.
+
+## Getting started
+
+### Configuration
+
+Ensure you are running within a python virtual environment, and ensure gcloud config project is set to your project. You can use the following command to set it:
+
+```shell
+gcloud config set project {PROJECT_ID}
+```
+
+### Environment variables:
+
+Set the following environment variables by updating the deployment file to run on a GKE cluster.
+1. Set the environment variables required to run the demo. These include:
+  
+  * `PROJECT`: The GCP project you're using.
+  
+  * `CKPT_DIR_PATH`: The full path of the directory in which to save checkpoints, in the format `gs://<bucket>/<directory>/`
+
+  * `WORLD_SIZE`: The number of nodes.
+
+  * `LAYER_SIZE`: The size of each layer.
```

Comment:
Is this really accurate that it's the size? I thought it was actually the count of layers, in which case the naming may need to be updated. In either case, can you add some more information about what this means?

Reply:
Yes it is accurate. This is layer_size. If we look at the code,  layer_size is number of neurons in each layer.  padding_size on the other hand decides the total number of layers. (But the name is still kept to padding_size since we are sort of padding it to the state_dict. 

Added some documentation around layer_size.